Fixed #15910 -- Displayed delete buttons for all admin inline formset rows.

arthurdejong opened the pull request:
I have updated the patch for #15910 to apply clean to current master, include a Selenium test and also update the minified file.

There is also a commit with a fix in a related test that is be more quickly exposed with the made changes.

All the tests (including the Selenium ones) pass.

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
Could you please check the JavaScript/flake8 errors?

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -750,9 +750,10 @@ def test_delete_stackeds(self):
-        for delete_link in self.selenium.find_elements_by_css_selector(
-                '%s .inline-deletelink' % inline_id):
-            delete_link.click()
+        self.selenium.find_element_by_css_selector(
```
This can be 1 line (we favor lines up to 119 characters when it helps readability)

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -854,6 +855,49 @@ def test_delete_inlines(self):
+    def test_gone_deletes(self):
+        """
+        Ensure that the delete buttons are also present in the form after
+        a form validation error. Bug #15910.
```
We don't really need to include the reference to the ticket except in obscure situations where the behavior could benefit from additional explanation that doesn't fit in a docstring. I don't think that's the case here.

timgraham commented on django/contrib/admin/static/admin/js/inlines.js:
```diff
@@ -117,6 +85,46 @@
                 });
+            }
+        });
+        // only show the add button if we are allowed to add more items,
```
Is all this code unchanged and just moved from above? (difficult to review)

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -854,6 +855,49 @@ def test_delete_inlines(self):
+        self.selenium.find_element_by_link_text('Add another Inner2').click()
+        self.assertEqual(len(self.selenium.find_elements_by_css_selector(
+            '#inner2_set-group div.dynamic-inner2_set')), 5)
+        self.assertEqual(len(self.selenium.find_elements_by_css_selector(
```
Please use this indent style:
```
self.assertEqual(
    len(self.selenium.find_elements_by_css_selector('form#holder2_form div.dynamic-inner2_set#inner2_set-0')
    1
)
```
(maybe a loop or helper function would make this less redundant and easier to read... or factoring "form#holder2_form div.dynamic-inner2_set#inner2_set-" into a constant so it's not repeated so much?)

timgraham commented:
Please uncheck "Patch needs improvement" after updating, thanks!

arthurdejong commented:
The test was more or less based on test_delete_inlines() and the code layout of that test was used. I've added commits to use a lambda function which is also used elsewhere in the same module. I've also updated the code style.

Please let me know if you prefer to have these as separate commits at this points or as a rebased change.

I'll address the Javascript changes separately because I found another issue related to the add link in the formset that could show up when form validation fails (found due to the Javascript test).

arthurdejong commented:
The Javascript changes are initially moving some code related to the delete link around, adding a condition/loop around it and adding row variables. This way, the delete link is added to all rows instead of using the add trigger to add the delete link. The delete link is also added to the hidden template form so it automatically shows up in the rows added by clicking the "Add another.." link.

I've just pushed changes that fix the issues found by the javascript test. I've made a few commits because it includes more code changes in the moved code and making one commit will make it more difficult to review. Please let me know if there is more I can do to make it easier to review.

The code change also exposes a bug in the showing/hiding of the "add another" link when using max_num: if you get a form validation error after the maximum number of inlines has been reached the "add another" link is not added to the form so cannot be shown if an inline is removed. The bug cannot be triggered without the change in this issue. Should this be made into a separte ticket and how should these changes be linked (I have a test and fix ready)?

timgraham commented:
Feel free to submit that fix as a separate commit on this pull request.

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -854,6 +853,47 @@ def test_delete_inlines(self):
+        a form validation error.
+        """
+        self.admin_login(username='super', password='secret')
+        self.selenium.get('%s%s' % (self.live_server_url,
```
I've started using this style: `self.selenium.get(self.live_server_url + reverse('admin:admin_inlines_holder2_add'))`

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -854,6 +853,47 @@ def test_delete_inlines(self):
+        self.selenium.get('%s%s' % (self.live_server_url,
+            reverse('admin:admin_inlines_holder2_add')))
+
+        inlines = lambda: self.selenium.find_elements_by_css_selector(
```
I think I'd name these like `get_inlines`, `get_add_another()`, etc.

timgraham commented:
Closing due to inactivity. Please send a new PR if you return to this.
